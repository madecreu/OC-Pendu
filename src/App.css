.App {
    text-align: center;
}

.word {
    letter-spacing: 0.3em;
}

p {
    letter-spacing: 0.5em;
}

button {
    margin: 0.3em;
}

input {
    opacity: 0;
}