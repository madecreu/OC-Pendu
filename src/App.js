import Keyboard from "./Keyboard";
import ListUsedLetters from "./ListUsedLetters";
import './App.css'
import randomWords from 'random-words'
import {useEffect, useState} from "react";

const computeDisplay = (word, usedLetters) => word.replace(/\w/g, (letter) => (usedLetters.includes(letter) ? letter : '_'));

const App = () => {
    const [word, setWord] = useState(randomWords().toUpperCase());
    const [usedLetters, setUsedLetters] = useState([]);
    const [count, setCount] = useState(0);
    const [end, setEnd] = useState(false);

    useEffect(() => {
        if (!computeDisplay(word, usedLetters).includes('_')) {
            setEnd(true);
        }
    }, [usedLetters, end, word]);


    const handleTry = (event) => {
        const triedLetter = event.currentTarget.value.toUpperCase();
        setUsedLetters((usedLetters) => [...usedLetters, triedLetter]);
        setCount((count) => count + 1);
    }

    const handleRestart = () => {
        setUsedLetters([]);
        setCount(0);
        setWord(randomWords().toUpperCase());
        setEnd(false);
    }

    return (
        <div className="App">
            <h1 className="word" name="word">{computeDisplay(word, usedLetters)}</h1>
            <h3>Nombres d'essais : {count}</h3>
            <div>
                {end
                    ? <button onClick={() => handleRestart()}>Restart</button>
                    : <Keyboard handleTry={handleTry} />
                }
            </div>
            <div>
                <h2>Lettres déjà essayées :</h2>
                <ListUsedLetters usedLetters={usedLetters}/>
            </div>
        </div>
    );
}

export default App;
