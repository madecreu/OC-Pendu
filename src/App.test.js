import randomWords from 'random-words';
import {render, fireEvent, screen} from '@testing-library/react'
import App from './App';

jest.mock('random-words');

const fireMultipleChanges = (items) => {
    for (let i = 0; i < items.length; i++) {
        fireEvent.change(screen.getByRole('textbox'), {target: {value: items[i]}})
    }
}

describe('<App />', () => {
    beforeEach(() => {
        randomWords.mockImplementationOnce(() => 'testing').mockImplementation(() => 'restarting');
        render(<App />);
    });

   describe('Initialization', () => {
      it ('should render a word', () => {
          expect(screen.getByRole('heading', {level: 1})).toHaveTextContent('_______');
          expect(screen.getByTestId('usedLetters')).toBeEmptyDOMElement();
      });
   });

   describe('Actions', () => {
      it('should change word displayed', () => {
          fireEvent.change(screen.getByRole('textbox'), {target: {value: 'E'}});
          expect(screen.getByRole('heading', {level: 1})).toHaveTextContent('_E_____');
          expect(screen.getByTestId('usedLetters')).toHaveTextContent(/^E$/);
          fireMultipleChanges(['T', 'S', 'I', 'N', 'G']);
          expect(screen.getByRole('heading', {level: 1})).toHaveTextContent('TESTING');
          expect(screen.getByTestId('usedLetters')).toHaveTextContent(/^ETSING$/);
      });

      it('should display wrong letters', () => {
          fireEvent.change(screen.getByRole('textbox'), {target: {value: 'A'}});
          expect(screen.getByRole('heading', {level: 1})).toHaveTextContent('_______');
          expect(screen.getByTestId('usedLetters')).toHaveTextContent(/^A$/);
      })

      it('should show restart button at the end', () => {
          expect(screen.queryByRole('button')).toBeNull();
          fireMultipleChanges(['T', 'E', 'S', 'I', 'N', 'G']);
          expect(screen.getByRole('button')).toHaveTextContent('Restart');
      });

      it('should restart game at the end', () => {
          fireMultipleChanges(['T', 'E', 'S', 'I', 'N', 'G']);
          fireEvent.click(screen.getByRole('button'));
          expect(screen.getByRole('heading', {level: 1})).toHaveTextContent('__________');
          expect(screen.queryByRole('button')).toBeNull();
          expect(screen.getByTestId('usedLetters')).toBeEmptyDOMElement();
      });
   });

   describe('Snapshot', () => {
      it('should match with previous Snapshot', () => {
          fireMultipleChanges(['T', 'E', 'I', 'A', 'V']);
          expect(render(<App />).baseElement).toMatchSnapshot();
      });
   });
});