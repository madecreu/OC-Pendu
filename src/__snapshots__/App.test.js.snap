// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Snapshot should match with previous Snapshot 1`] = `
<body>
  <div>
    <div
      class="App"
    >
      <h1
        class="word"
        name="word"
      >
        TE_TI__
      </h1>
      <h3>
        Nombres d'essais : 
        5
      </h3>
      <div>
        <div>
          <input
            type="text"
            value=""
          />
        </div>
      </div>
      <div>
        <h2>
          Lettres déjà essayées :
        </h2>
        <div>
          <p
            data-testid="usedLetters"
          >
            T
            E
            I
            A
            V
          </p>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      class="App"
    >
      <h1
        class="word"
        name="word"
      >
        __________
      </h1>
      <h3>
        Nombres d'essais : 
        0
      </h3>
      <div>
        <div>
          <input
            type="text"
            value=""
          />
        </div>
      </div>
      <div>
        <h2>
          Lettres déjà essayées :
        </h2>
        <div>
          <p
            data-testid="usedLetters"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
